feat: add GitFile Component

raphaelchristi opened the pull request:
This pull request introduces a new GitFileComponent that allows users to analyze Git repositories and retrieve the content of selected files from specified branches. The component provides the following features:

- Repository URL input for connecting to Git repositories
- Branch selection dropdown that dynamically fetches available branches
- File selection interface to choose specific files from the repository
- File content retrieval functionality with binary file detection
- Error handling for various Git operations

raphaelchristi committed:
feat: add GitFileComponent

raphaelchristi committed:
fix: correct linting and formatting issues in GitFileComponent

raphaelchristi committed:
fix: add GitFileComponent to git module exports

raphaelchristi committed:
fix: add GitFileComponent to git module exports

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

ogabrielluiz reviewed (changes requested):
Please take a look at the other Git component to check an example on how to use anyio.Path

raphaelchristi committed:
refactor(GitFile): implement anyio.Path for async file operations

- Replace pathlib.Path with anyio.Path
- Convert file operations to async (open, exists, stat, rglob)
- Update file traversal to use async pattern
- Keep async context manager for cleanup

ogabrielluiz commented on src/backend/base/langflow/components/git/gitfile.py:
```diff
@@ -0,0 +1,186 @@
+        Output(display_name="Files Content", name="files_content", method="get_files_content"),
+    ]
+
+    @asynccontextmanager
+    async def temp_git_repo(self):
+        """Async context manager for temporary git repository cloning."""
+        temp_dir = tempfile.mkdtemp()
+        try:
+            yield temp_dir
+        finally:
+            await asyncio.get_event_loop().run_in_executor(None, lambda: shutil.rmtree(temp_dir, ignore_errors=True))
```
```suggestion
    @asynccontextmanager
    async def temp_git_repo(self):
        """Context manager for handling temporary clone directory."""
        temp_dir = None
        try:
            temp_dir = tempfile.mkdtemp(prefix="langflow_clone_")
            yield temp_dir
        finally:
            if temp_dir:
                await anyio.Path(temp_dir).rmdir()
```

ogabrielluiz commented on src/backend/base/langflow/components/git/gitfile.py:
```diff
@@ -0,0 +1,186 @@
+        ):
+            try:
+                async with self.temp_git_repo() as temp_dir:
+                    await asyncio.get_event_loop().run_in_executor(
+                        None,
+                        lambda: git.Repo.clone_from(
+                            self.repository_url, temp_dir, branch=self.branch, depth=1, single_branch=True
+                        ),
+                    )
```
This will not work. Use this instead:
```suggestion
                    await asyncio.to_thread(
                git.Repo.clone_from,
                self.repository_url,
                temp_dir,
                branch=self.branch,
                depth=1,
                single_branch=True,
            )
```

ogabrielluiz commented on src/backend/base/langflow/components/git/gitfile.py:
```diff
@@ -0,0 +1,186 @@
+        return build_config
+
+    async def get_files_content(self) -> list[Data]:
+        if not self.repository_url:
+            return [Data(data={"error": "Please enter a repository URL"})]
+        if not self.branch or self.branch == "Enter repository URL first":
+            return [Data(data={"error": "Please select a branch"})]
+        if not self.selected_files:
+            return [Data(data={"error": "Please select at least one file"})]
```
These should be exceptions.

ogabrielluiz commented on src/backend/base/langflow/components/git/gitfile.py:
```diff
@@ -0,0 +1,186 @@
+
+        try:
+            async with self.temp_git_repo() as temp_dir:
+                await asyncio.get_event_loop().run_in_executor(
+                    None,
+                    lambda: git.Repo.clone_from(
+                        self.repository_url, temp_dir, branch=self.branch, depth=1, single_branch=True
+                    ),
+                )
```
Never use asyncio.get_event_loop() because there may not be a running event loop which will raise an error. Always use `asyncio.to_thread`.

ogabrielluiz reviewed (changes requested):
Thanks for this, @raphaelchristi !

raphaelchristi committed:
fix: improve GitFile component handling

- Replace asyncio.get_event_loop() with asyncio.to_thread()
- Add proper error handling with GitFileError
- Improve temporary directory cleanup
- Make required fields explicit
- Add type hints and docstrings
- Enhance error messages and logging

Fixes #5458

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

raphaelchristi committed:
Revert "fix: improve GitFile component handling"

This reverts commit 1e0930708a811012cdd40f3f32123ccfafcb1e26.

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

raphaelchristi committed:
fix: address code review feedback for GitFile component

- Replace temp dir cleanup with direct shutil.rmtree
- Use appropriate Git-specific exceptions
- Directly propagate GitCommandError
- Improve error messaging

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

raphaelchristi committed:
Fix linting and formatting issues

raphaelchristi commented:
Thanks for the thorough review @ogabrielluiz ! I've implemented your suggested changes:

1. Replaced all `asyncio.get_event_loop()` calls with `asyncio.to_thread()`
2. Improved temp directory handling with proper cleanup
3. Added proper error handling with a custom `GitFileError` class
4. Fixed other code quality issues

Please let me know if you'd like me to make any additional changes.

anovazzi1 committed:
Refactor: Update EditNodeComponent to hide table options and block hide (#5513)

ogabrielluiz committed:
refactor(GitFile): enhance async handling and temporary directory management

- Introduced `anyio` for improved async file operations.
- Updated temporary directory creation and cleanup logic for better error handling and efficiency.
- Added a test case handling mechanism in the `__call__` method.
- Improved binary file detection and content reading using `aiofiles`.
- Enhanced error logging during temporary directory cleanup.

These changes aim to streamline the GitFileComponent's functionality and improve overall performance.

ogabrielluiz committed:
test: add unit tests for GitFileComponent

- Introduced comprehensive unit tests for the GitFileComponent to validate its behavior.
- Added tests for handling missing repository URL, branch, and selected files.
- Implemented tests for binary and text file detection.
- Created tests for retrieving branches and successfully getting file contents.
- Utilized mocking to simulate Git operations and file handling.

These tests aim to ensure the reliability and correctness of the GitFileComponent's functionality.

edwinjosechittilappilly commented:
Experiencing issues when loading a Git repository. I also suggest adding real_time_refresh=True, so that as soon as the Git URL is updated, the branch and the subsequent fields are refreshed automatically instead of having to refresh each field manually.


Issue Faced:
<img width="337" alt="Screenshot 2025-01-16 at 12 52 21 PM" src="https://github.com/user-attachments/assets/c3c0f866-6d8f-4b6c-b8c1-ed58dc17f042" />

edwinjosechittilappilly reviewed (changes requested):
Utilise real_time_refresh

Also utilise update_build_config to update the fields data instead of multiple refresh buttons.

raphaelchristi committed:
fix: unify Git component imports

raphaelchristi committed:
Merge branch 'main' into feature/git-file-component

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes (attempt 2/3)

raphaelchristi committed:
Merge branch 'main' into feature/git-file-component

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

raphaelchristi committed:
Merge branch 'main' into feature/git-file-component

cbornet commented on src/backend/base/langflow/components/git/gitfile.py:
```diff
@@ -0,0 +1,230 @@
+    ]
+
+    def __call__(self, *args: Any, **kwargs: Any) -> Any:
+        """Override to handle test cases."""
```
Can we avoid test code leaking into main code ?
Is it possible to mock inside the test instead ?

raphaelchristi committed:
Merge branch 'main' into feature/git-file-component

edwinjosechittilappilly committed:
Merge branch 'main' into feature/git-file-component

edwinjosechittilappilly commented:
@raphaelchristi  Is it ready for review?
